### ISSUE TITLE
sigmoidCrossEntropy is not differentiable

### ISSUE THREAD

**Issue by tanmayb123** [tensorflow/swift-apis#1038]
I'm working on an example of Swift for TensorFlow and, for some reason, `sigmoidCrossEntropy` doesn't seem to be differentiable.

Code:
```
@differentiable
func unetWorkerLoss(model: UNet, input: Tensor<Float>, output: Tensor<Float>) -> Tensor<Float> {
    sigmoidCrossEntropy(logits: model(input).reshaped(to: [-1, 256 * 256]), labels: output.reshaped(to: [-1, 256 * 256]))
}
```

Error:
```
/home/tanmay/swift-models/UNet/main.swift:77:5: error: expression is not differentiable
    sigmoidCrossEntropy(logits: model(input).reshaped(to: [-1, 256 * 256]), labels: output.reshaped(to: [-1, 256 * 256]))
    ^
/home/tanmay/swift-models/UNet/main.swift:77:5: note: cannot differentiate functions that have not been marked '@differentiable' and that are defined in other files
    sigmoidCrossEntropy(logits: model(input).reshaped(to: [-1, 256 * 256]), labels: output.reshaped(to: [-1, 256 * 256]))
```

**Comment by tanmayb123** [comment]
However, when I do copy in the function directly from swift-apis with no modifications, and change the function names:

```
@differentiable(wrt: logits)
public func tbSigmoidCrossEntropy<Scalar: TensorFlowFloatingPoint>(
  logits: Tensor<Scalar>,
  labels: Tensor<Scalar>,
  reduction: @differentiable (Tensor<Scalar>) -> Tensor<Scalar> = _mean
) -> Tensor<Scalar> {
  let device = logits.device
  // This numerically stable implementation is based on the TensorFlow Python API.
  let maxLogitsWithZero = max(logits, Tensor(0, on: device))
  let negAbsLogits = max(logits, -logits)  // Custom `abs` to compute gradients at `0`.
  return reduction(maxLogitsWithZero - logits * labels + log1p(exp(-negAbsLogits)))
}

@differentiable
func unetWorkerLoss(model: UNet, input: Tensor<Float>, output: Tensor<Float>) -> Tensor<Float> {
    tbSigmoidCrossEntropy(logits: model(input).reshaped(to: [-1, 256 * 256]), labels: output.reshaped(to: [-1, 256 * 256]))
}
```

It works.

**Comment by mikowals** [comment]
I don't think the original report is a bug though the error message could be improved to say that `sigmoidCrossEntropy` is only differentiable with respect to `logits`.   `unetWorkerLoss` attempts to differentiate with respect to `labels` also.  Changing `@differentiable` to `@differentiable(wrt: input)` in the code above removes the error.  Full working example:

    import TensorFlow

    struct UNet: Layer {
      var x = Tensor<Float>(1)
      
      func callAsFunction(_ input: Tensor<Float>) -> Tensor<Float> {
         x + input
      }
    }

    @differentiable(wrt: input)
    func unetWorkerLoss(model: UNet, input: Tensor<Float>, output: Tensor<Float>) -> Tensor<Float> {
      sigmoidCrossEntropy(logits: model(input).reshaped(to: [-1, 256 * 256]), labels: output.reshaped(to: [-1, 256 * 256]))
    }

It is unexpected (and maybe inefficient) that the workaround of rewriting `sigmoidCrossEntropy` works but it appears `@differentiable` is not needed (and functions are always differentiable with respect to all inputs) for differentiation within the same file.  For example, this code compiles and executes without error:

    import TensorFlow

    //@differentiable(wrt: x) // Code still works if this is uncommented.
    func addXY(x: Tensor<Float>, y: Tensor<Float>) -> Tensor<Float> {
       x + y
    }

    // No need for @differentiable here either.
    func wrapAddXY(x: Tensor<Float>, y: Tensor<Float>) -> Tensor<Float> {
       return addXY(x: x, y: y)
    }

    let grad = gradient(at: Tensor<Float>(1), Tensor<Float>(2), in: wrapAddXY)
    print(grad) // (1.0, 1.0)

**Comment by dan-zheng** [comment]
> I don't think the original report is a bug though the error message could be improved to say that `sigmoidCrossEntropy` is only differentiable with respect to `logits`.

Yes, this is accurate: the current diagnostic messages are misleading. Instead, they should read something like:

```
/home/tanmay/swift-models/UNet/main.swift:77:5: error: expression is not differentiable
    sigmoidCrossEntropy(logits: model(input).reshaped(to: [-1, 256 * 256]), labels: output.reshaped(to: [-1, 256 * 256]))
    ^
/home/tanmay/swift-models/UNet/main.swift:77:5: note: 'foo' is marked '@differentiable' only with respect to a smaller subset of arguments: '@differentiable(wrt: logits)'
    sigmoidCrossEntropy(logits: model(input).reshaped(to: [-1, 256 * 256]), labels: output.reshaped(to: [-1, 256 * 256]))
```

SR-13224 tracks improving this diagnostic, we'll look into it soon!

**Comment by marcrasi** [comment]
Closing because `sigmoidCrossEntropy` is sufficiently differentiable, and I made a comment in https://github.com/tensorflow/swift-models/pull/623 explaining how to make it work.
